### PR TITLE
Avoid modification of input dataset.

### PR DESCRIPTION
**Pull request by davidlehn**
Fixes https://github.com/digitalbazaar/rdf-canonize/issues/20

Benchmark testing on a noisy machine made this look neutral.  Would be good to benchmark on more real data to ensure this isn't a regression.

### CODE REVIEW

File: lib/NQuads.js [hunk 1]
```diff
@@ -194,18 +194,16 @@ module.exports = class NQuads {
...
-   * Converts an RDF quad to an N-Quad string (a single quad).
+   * Converts RDF quad components to an N-Quad string (a single quad).
    *
-   * @param quad the RDF quad convert.
+   * @param {Object} s - N-Quad subject component.
+   * @param {Object} p - N-Quad predicate component.
+   * @param {Object} o - N-Quad object component.
+   * @param {Object} g - N-Quad graph component.
    *
    * @return the N-Quad string.
```
Comment (mattcollier) [inline lib/NQuads.js]:
```suggestion
   * @return {string} the N-Quad.
```

**Comment by mattcollier** [comment]
@davidlehn I don't know what your existing benchmarking was, but this one might allow you build some more varied test cases https://github.com/digitalbazaar/rdf-canonize-native/blob/master/benchmark/benchmark2.js

**Comment by davidlehn** [comment]
@mattcollier I was using the regular benchmark.js one.

**Comment by davidlehn** [comment]
Need to check if the previous code was optimizing a particular graph case where the same blank node was visited many times and overwriting the input caused it to skip future gen id calls.  @dlongley mentioned this, but it may be for a variant of the current code.  It looks like it's just going over all the quads, and they are not linked in memory, so each one is processed once.

File: lib/NQuads.js [hunk 2]
```diff
@@ -194,18 +194,16 @@ module.exports = class NQuads {
...
-   * @return the N-Quad string.
+   * @return {string} the N-Quad.
    */
-  static serializeQuad(quad) {
-    const s = quad.subject;
-    const p = quad.predicate;
-    const o = quad.object;
-    const g = quad.graph;
-
+  static serializeQuadComponents(s, p, o, g) {
```
Comment (dlongley) [inline lib/NQuads.js]: Could use destructuring here instead: `({subject: s, predicate: p, object: o, graph: g})`. This is internal anyway, so can always change later.

File: lib/URDNA2015.js [hunk 3]
```diff
@@ -209,14 +209,14 @@ module.exports = class URDNA2015 extends AsyncAlgorithm {
               // blank node identifiers using the canonical identifiers
               // previously issued by canonical issuer.
               // Note: We optimize away the copy here.
```
Comment (dlongley) [inline lib/URDNA2015.js]: If we don't optimize away the copy anymore, we should remove this comment.

**Comment by davidlehn** [comment]
This PR was a bit stale.  I forget why it was never merged.  Some of the [old changes](https://github.com/digitalbazaar/rdf-canonize/tree/fix-input-changes-old) no longer applied.  And I don't think the original issue problems still exist.  But this does add tests to make sure.  I rebased on recent code.  I think there may still be value in micro-optimizing away copies and avoiding extra temporary data structure manipulation for local hot paths.

**Review by dlongley — APPROVED** [review]

**Pull request merged**